Add RPM package build to travis ci script

andreas-bok-sociomantic opened the pull request:
Add a Docker build file and set up the
travis config to build the image

Run the container and build the rpm package

mathias-lang-sociomantic commented on docker/Dockerfile:
```diff
@@ -0,0 +1,11 @@
+# Docker build file to create a centos image for building rpm packages
+
+FROM centos:7
+MAINTAINER andreas-bok-sociomantic <andreas-bok@sociomantic.com>
```
Isn't it `andreas.bok` ? Maybe we could have a public address for the project(s) ?

mathias-lang-sociomantic commented on docker/Dockerfile:
```diff
@@ -0,0 +1,11 @@
+RUN yum install -y rpm-build zlib-devel libtool openssl \
+    openssl-devel automake autoconf gcc-c++ && \
+    yum clean all && \
+    rm -r -f /var/cache/*
```
Y ?

andreas-bok-sociomantic replied:
You you mean `yum install -Y`? 

if so, **no**: 

```
yum help install
...
  -y, --assumeyes       Automatically answer yes for all question
```

mathias-lang-sociomantic replied:
No, I mean "Why" :P
Why do you `rm -rf /var/cache/*` ?

andreas-bok-sociomantic replied:
Ah. See your point since the container is transient... :)

andreas-bok-sociomantic replied:
Ah. See your point since the container is transient... :)

`yum clean all && \` can be omitted as well

mathias-lang-sociomantic commented on rpm/include.am:
```diff
@@ -29,6 +29,14 @@ clean-rpm:
 rpm: rpm-build
 
+docker-rpm:
+	@sed -i 's/%{__make} check/#%{__make} check/g' "rpm/spec.in"
```
This could use a comment as well. Why are we disabling this check ?

andreas-bok-sociomantic replied:
Basically to avoid setting up a MySQL server in the docker container. The check test is already done [here](https://github.com/andreas-bok-sociomantic/libdrizzle-redux/blob/dc05295f05bdd3c3cd8b572e14e3ce7bd441f4ce/.travis.yml#L65)

And it reduces build time

mathias-lang-sociomantic commented on .travis.yml:
```diff
@@ -63,4 +64,6 @@ script:
   - make check
   - make deb
+  - docker build -f docker/Dockerfile -t centos:7 .
+  - docker run --rm -v $PWD:/home centos:7 make docker-rpm
```
Sounds like this could be hugely simplified if you used a [build matrix](https://docs.travis-ci.com/user/customizing-the-build#Build-Matrix).

andreas-bok-sociomantic replied:
Using build matrix is already part of [#162](https://github.com/sociomantic-tsunami/libdrizzle-redux/pull/162/files#diff-354f30a63fb0907d4ad57269548329e3R51)

mathias-lang-sociomantic replied:
Yeah I reviewed this one later. IMO the approach you have here is complicated by the fact you are not using build matrix. I would strongly recommend to go with the build matrix, then update this P.R. based on it.

andreas-bok-sociomantic replied:
>  IMO the approach you have here is complicated by the fact you are not using build matrix. I would strongly recommend to go with the build matrix, then update this P.R. based on it.

Wrt. the RPM build to the travis script I am not sure much is gained by using the matrix. Building the Dockerfile and running the container still needs to be done explicitly in the `script` section. 
I can integrate the matrix into this PR and move the RPM build commands to the aux script as I have done in #162, so it is prepared for merging with #162 but again IMHO solely wrt to the RPM build it won't make much of a difference

andreas-bok-sociomantic commented:
Updated based on comments from @mathias-lang-sociomantic 

```diff
diff --git a/.travis.yml b/.travis.yml
index c9b909d..b2e3839 100644
--- a/.travis.yml
+++ b/.travis.yml
@@ -66,4 +66,3 @@ script:
   - make deb
   - docker build -f docker/Dockerfile -t centos:7 .
   - docker run --rm -v $PWD:/home centos:7 make docker-rpm
-
diff --git a/docker/Dockerfile b/docker/Dockerfile
index 2d7c21a..fe113b3 100644
--- a/docker/Dockerfile
+++ b/docker/Dockerfile
@@ -1,7 +1,7 @@
 # Docker build file to create a centos image for building rpm packages
 
 FROM centos:7
-MAINTAINER andreas-bok-sociomantic <andreas-bok@sociomantic.com>
+MAINTAINER andreas-bok-sociomantic <andreas.bok@sociomantic.com>
 
 RUN yum install -y rpm-build zlib-devel libtool openssl \
     openssl-devel automake autoconf gcc-c++ && \
diff --git a/rpm/include.am b/rpm/include.am
index cd160a5..c1f2f7f 100644
--- a/rpm/include.am
+++ b/rpm/include.am
@@ -35,6 +35,10 @@ docker-rpm:
 	$(SHELL) ./configure
 	$(MAKE) rpm
 	$(MAKE) clean-rpm
+	# Disable running the unittests in the %check section. The check is done
+	# previously and it is obsolete for checking that the RPM package is built
+	# without errors. Further more it reduces the build time since it isn't
+	# necessary to set up a MySQL server in the docker container
 	@sed -i 's/#%{__make} check/%{__make} check/g' "rpm/spec.in"
 
 release: rpm-sign

```

andreas-bok-sociomantic committed:
Add Dockerfile for building rpm package

The Dockerfile will build a centos7 image with MySQL 5.6
and required libraries/apps for building libdrizzle-redux

andreas-bok-sociomantic committed:
Add RPM pkg build to ci travis build script

Refactor the .travis.yml script to call functions
in a separate bash script for each build step

Use a build matrix to define the configurations for
builds

andreas-bok-sociomantic commented:
Updated travis config file to use 'build matrix' to configure jobs.

daniel-zullo commented on docker/Dockerfile:
```diff
@@ -0,0 +1,34 @@
+# Change user to `root` and install required packages to build libdrizzle-redux.
+USER root
+RUN yum install -y deltarpm
+RUN yum install -y rpm-build zlib-devel libtool openssl \
```
I guess you could have included `deltarpm` here and delete the line above. Or am I missing something?

andreas-bok-sociomantic replied:
What do you mean?
I wanted to install deltarpm first, so the less stuff has to be downloaded when installing the next packages

daniel-zullo replied:
I meant that you could have merged the two commands into one.

daniel-zullo replied:
> Yes, but the purpose of first installing deltarpm is that yum uses it to only download the diff of a update. So it needs that app to be installed to perform that action when installing the next packages

ok, got it

daniel-zullo commented on travis_configure.sh:
```diff
@@ -0,0 +1,96 @@
+#!/usr/bin/env bash
```
Usually `sh` is used for package configuration/installation. Is there any special reason why `bash` is needed?

andreas-bok-sociomantic replied:
No particular reason, will change.

daniel-zullo commented on travis_configure.sh:
```diff
@@ -0,0 +1,96 @@
+function enable_mysqlbinlog
+{
+    sudo sed -i -e 's/^# *server.id.*$/server_id=1/g' $1
+    sudo sed -i -e 's/^# *log.bin.*$/log_bin=mysql-bin/g' $1
```
Do you really need `sudo` here?

andreas-bok-sociomantic replied:
I will try to remove and see if it passes

andreas-bok-sociomantic replied:
Nope, sudo is required to write to the root folder `/etc`. it is required for calling `service`

daniel-zullo commented on travis_configure.sh:
```diff
@@ -0,0 +1,96 @@
+            #  1. Enable replication in mysql config file
+            #  2. Restart MySQL server
+            enable_mysqlbinlog "/etc/mysql/my.cnf"
+            sudo service mysql restart
```
do you also need to use `sudo` here?

andreas-bok-sociomantic commented:
> I meant that you could have merged the two commands into one.

Yes, but the purpose of first installing deltarpm is that yum uses it to only download the diff of a update. So it needs that app to be installed to perform that action when installing the next packages

andreas-bok-sociomantic commented:
Updated based on comments from @daniel-zullo-sociomantic 
Feel free to merge, if there are no further comments

daniel-zullo commented on travis_configure.sh:
```diff
@@ -0,0 +1,99 @@
+#!/bin/bash
+# Script for setting up OS specific CI test environment on travis-ci.org
+#
+# Copyright (C) 2016 Andreas Bok Andersen <andreas.bok@sociomantic.com>
```
2017

andreas-bok-sociomantic replied:
True, fixed

daniel-zullo commented on travis_configure.sh:
```diff
@@ -0,0 +1,99 @@
+#!/bin/bash
```
what about `#!/bin/sh`?

andreas-bok-sociomantic replied:
Sorry made a comment but somehow didn't post it. I tried with `sh` but some of the syntax is  not compatible. Furthermore `travis-ci` uses bash scripts in their examples in their github repo. So I will stick with `bash`

daniel-zullo reviewed (commented):
I haven't finished my reviewed yet

daniel-zullo commented on travis_configure.sh:
```diff
@@ -0,0 +1,99 @@
+{
+    if [[ "$TRAVIS_OS_NAME" == "linux" ]]; then
+        if [[ "$DIST_PACKAGE_TARGET" == "DEB" ]]; then
+            gem install fpm
```
I wonder if this command needs `sudo` then.

andreas-bok-sociomantic replied:
It doesn't. But I will try to add a printout of the gem env variables to see where gems are installed into. Might install it into $HOME/.gems which wouldn't require sudo

andreas-bok-sociomantic replied:
Ok checked the docs for travis. They don't use sudo for their examples of installing gems, 
https://docs.travis-ci.com/user/languages/ruby/

daniel-zullo replied:
fair enough

daniel-zullo reviewed (commented):
I only have a question and LGTM.

@mathias-lang-sociomantic any further comments?

andreas-bok-sociomantic commented:
ping @mathias-lang-sociomantic

mathias-lang-sociomantic commented on travis_configure.sh:
```diff
@@ -0,0 +1,99 @@
+#!/bin/bash
+# Script for setting up OS specific CI test environment on travis-ci.org
+#
+# Copyright (C) 2017 Andreas Bok Andersen <andreas.bok@sociomantic.com>
```
Sociomantic labs GmbH

mathias-lang-sociomantic commented on travis_configure.sh:
```diff
@@ -0,0 +1,99 @@
+            mysql -u root -e "SET PASSWORD FOR 'root'@'localhost' = PASSWORD('')"
+        elif [[ "$DIST_PACKAGE_TARGET" == "RPM" ]]; then
+            # build a docker image with Centos 7 and MySQL
+            docker build -t centos7_mysql56 -f docker/Dockerfile .
```
else fail ?

andreas-bok-sociomantic replied:
Hmm, yes. I gues it could be good to add a notice about an invalid build configuration

andreas-bok-sociomantic replied:
But it is not in all cases where we wan't to exit with an error. the `before_install` script should only handle the linux/deb case

mathias-lang-sociomantic replied:
In this case make it explicit with a comment. The underlying concern was that, if you want to add an architecture / test, you don't want to miss out on the configuration somewhere (although it will most likely be visible if you do :) )

andreas-bok-sociomantic replied:
Sure, the build matrix does expand exponentially when whenever a new build setting is added. So handling invalid cases would be helpful

mathias-lang-sociomantic commented on travis_configure.sh:
```diff
@@ -0,0 +1,99 @@
+        elif [[ "$DIST_PACKAGE_TARGET" == "RPM" ]]; then
+            docker run -d --name db_container centos7_mysql56
+            docker exec -u root db_container sh -c \
+                "autoreconf -fi && ./configure && make rpm"
```
Ditto else fail

mathias-lang-sociomantic commented on travis_configure.sh:
```diff
@@ -0,0 +1,99 @@
+    if [[ "$TRAVIS_OS_NAME" == "linux" ]]; then
+        if [[ "$DIST_PACKAGE_TARGET" == "DEB" ]]; then
+            gem install fpm
+        fi
```
fpm is here by default on rpm or not used ?

andreas-bok-sociomantic replied:
**fpm** is only used to build the deb package. The **rpm** package is built on the docker image. 
Just to avoid installing unnecessary apps.

andreas-bok-sociomantic commented:
Addressed comments by @mathias-lang-sociomantic

mathias-lang-sociomantic reviewed (approved):
LGTM
